tutorial/book/src/dynamic/push_constants.md

```diff
@@ -57,29 +57,34 @@ Remove `model` from the `UniformBufferObject` struct since we will be specifying
 #[repr(C)]
 #[derive(Copy, Clone, Debug)]
 struct UniformBufferObject {
-    view: glm::Mat4,
-    proj: glm::Mat4,
+    view: Mat4,
+    proj: Mat4,
 }
 ```
 
 Also remove `model` from the `App::update_command_buffers` method.
 
 ```rust,noplaypen
-let view = glm::look_at(
-    &glm::vec3(2.0, 2.0, 2.0),
-    &glm::vec3(0.0, 0.0, 0.0),
-    &glm::vec3(0.0, 0.0, 1.0),
+let view = Mat4::look_at_rh(
+    point3(2.0, 2.0, 2.0),
+    point3(0.0, 0.0, 0.0),
+    vec3(0.0, 0.0, 1.0),
 );
 
-let mut proj = glm::perspective_rh_zo(
+let correction = Mat4::new(
+    1.0,  0.0,       0.0,       0.0,
+    0.0, -1.0,       0.0,       0.0,
+    0.0,  0.0, 1.0 / 2.0, 1.0 / 2.0,
+    0.0,  0.0,       0.0,       1.0,
+);
+
+let proj = correction * cgmath::perspective(
+    Deg(45.0),
```

Comment:
Based on running some of the examples, the position of the camera seems to have changed starting when this new correction matrix is introduced. I'd prefer that things stay the same since the screenshots in the tutorial reflect the old way things looked.

In any case, I will take a deeper look at all this matrix stuff. I'd prefer not to merge this until I understand what this is doing (I took a peek at https://matthewwellings.com/blog/the-new-vulkan-coordinate-system but that will take more than a minute or two to understand).

Reply:
Sorry for the late response, had COVID last week. I made some experiment with MATLAB, and I think I can explain this. Let's take a look at this step by step.

First, according to the definition of [`perspective_rh_no`](https://docs.rs/nalgebra-glm/latest/src/nalgebra_glm/ext/matrix_clip_space.rs.html#656-680) and [`perspective_rh_zo`](https://docs.rs/nalgebra-glm/latest/src/nalgebra_glm/ext/matrix_clip_space.rs.html#693-717), we can write the following two perspective matrix definitions:

![图片](https://github.com/KyleMayes/vulkanalia/assets/29348140/72f52333-6b76-4128-aaea-9414210ac418)

Then, since we're going to flip the Y axis, we flip the `(1, 1)` element of `perspective_rh_zo`:

![图片](https://github.com/KyleMayes/vulkanalia/assets/29348140/124c30c3-0108-4fdf-9f03-31e0da8d413c)

To get the transformation matrix that could transform `perspective_rh_no` to `perspective_rh_zo_y_flipped`, *matrix division* can be used:

![图片](https://github.com/KyleMayes/vulkanalia/assets/29348140/5c60f5bc-befa-4a91-8436-b76d9dc67374)

And that's the matrix mentioned in https://matthewwellings.com/blog/the-new-vulkan-coordinate-system. Premultiplying `perspective_rh_no` with that matrix would yield `perspective_rh_zo_y_flipped`.

In depth, principally, this matrix can be thought as three separate transformations:

![图片](https://github.com/KyleMayes/vulkanalia/assets/29348140/3cdf0117-a956-4405-b5fd-26243c4ab503)

Then

![图片](https://github.com/KyleMayes/vulkanalia/assets/29348140/c77cfd84-51f0-4fd7-9b43-358ebcc6385d)

------

The reason why the position of camera changed is that I made a mistake when using `Mat4::new`, actually it should be

```rust
// Mat4::new constructs matrix in such a way that rows and columns seems "transposed"
let correction = Mat4::new(
    1.0,  0.0,       0.0, 0.0,
    0.0, -1.0,       0.0, 0.0,
    0.0,  0.0, 1.0 / 2.0, 0.0,
    0.0,  0.0, 1.0 / 2.0, 1.0,
);
```

Then the correction matrix should function properly. I've tested this with `26_depth_buffering.rs`

![图片](https://github.com/KyleMayes/vulkanalia/assets/29348140/075a0d42-a932-451c-8654-d09220f0b972)

(on the left is the code using `nalgebra-glm` on master branch, on the right is the code using `cgmath` + correction matrix)

I'd fix other correction matrices in the code then.